From ec182cc5529514dbd0362b6022daa7b3e2ca046d Mon Sep 17 00:00:00 2001
From: PulkoMandy <pulkomandy@pulkomandy.tk>
Date: Tue, 18 Aug 2026 22:04:44 +0200
Subject: Fix build on Haiku

- A few conflicting macros
- Both U-Boot and Haiku provide an image.h include, make sure the right
  one is used

diff --git a/arch/sandbox/include/asm/types.h b/arch/sandbox/include/asm/types.h
index 5f4b649e..ff8cf85c 100644
--- a/arch/sandbox/include/asm/types.h
+++ b/arch/sandbox/include/asm/types.h
@@ -8,7 +8,9 @@
 
 #include <asm-generic/int-ll64.h>
 
+#ifndef __HAIKU__
 typedef unsigned short umode_t;
+#endif
 
 /*
  * These aren't exported outside the kernel to avoid name space clashes
diff --git a/include/u-boot/ecdsa.h b/include/u-boot/ecdsa.h
index f0ac0f32..e72088d2 100644
--- a/include/u-boot/ecdsa.h
+++ b/include/u-boot/ecdsa.h
@@ -7,7 +7,7 @@
 #define _ECDSA_H
 
 #include <errno.h>
-#include <image.h>
+#include <../include/image.h>
 
 /**
  * crypto_algo API impementation for ECDSA;
diff --git a/lib/aes/aes-decrypt.c b/lib/aes/aes-decrypt.c
index 741102a4..00660505 100644
--- a/lib/aes/aes-decrypt.c
+++ b/lib/aes/aes-decrypt.c
@@ -6,7 +6,7 @@
 #ifndef USE_HOSTCC
 #include <malloc.h>
 #endif
-#include <image.h>
+#include <../include/image.h>
 #include <uboot_aes.h>
 
 int image_aes_decrypt(struct image_cipher_info *info,
diff --git a/lib/hash-checksum.c b/lib/hash-checksum.c
index 1970a741..74ab1fee 100644
--- a/lib/hash-checksum.c
+++ b/lib/hash-checksum.c
@@ -13,7 +13,7 @@
 #include "fdt_host.h"
 #endif
 #include <hash.h>
-#include <image.h>
+#include <../include/image.h>
 
 int hash_calculate(const char *name,
 		    const struct image_region *region,
diff --git a/tools/imagetool.h b/tools/imagetool.h
index d0e7d6d5..98f57a2e 100644
--- a/tools/imagetool.h
+++ b/tools/imagetool.h
@@ -15,18 +15,20 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <sys/param.h>
 #include <sys/stat.h>
 #include <sys/types.h>
 #include <time.h>
 #include <unistd.h>
 #include <u-boot/sha1.h>
 
-#include <image.h>
+#include <../include/image.h>
 
 #include "fdt_host.h"
 
 #define ARRAY_SIZE(x)		(sizeof(x) / sizeof((x)[0]))
 
+#undef ALIGN
 #define __ALIGN_MASK(x, mask)	(((x) + (mask)) & ~(mask))
 #define ALIGN(x, a)		__ALIGN_MASK((x), (typeof(x))(a) - 1)
 
diff --git a/tools/preload_check_sign.c b/tools/preload_check_sign.c
index ebead459..6a791c66 100644
--- a/tools/preload_check_sign.c
+++ b/tools/preload_check_sign.c
@@ -13,7 +13,7 @@
 #include <openssl/pem.h>
 #include <openssl/evp.h>
 #include <openssl/err.h>
-#include <image.h>
+#include <../include/image.h>
 
 extern void image_pre_load_sig_set_info(struct image_sig_info *info);
 extern int image_pre_load_sig(ulong addr);
diff --git a/tools/renesas_spkgimage.c b/tools/renesas_spkgimage.c
index ce3b2b28..9a2353c4 100644
--- a/tools/renesas_spkgimage.c
+++ b/tools/renesas_spkgimage.c
@@ -15,6 +15,8 @@
 #include <u-boot/crc.h>
 #include "renesas_spkgimage.h"
 
+#undef roundup
+
 /* Note: the ordering of the bitfields does not matter */
 struct config_file {
 	unsigned int version:1;
-- 
2.54.0

